# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
#Importing system-level tasks and data manipulation for processing data using panda
import sys
import pandas as pd

In [2]:
# My folder is located at my desktop in a folder, so I will add the folder containing the src module to Python's path
sys.path.append('/Users/<your-username>/Desktop/Group 8 - 62444 projects')


Importing the data from the Google drive file

In [3]:
import pandas as pd

#Here the data is loaded from Google Drive as provided:
def load_data_from_google_drive(url):
    """
    Converts a Google Drive link into a downloadable link and loads it as a pandas DataFrame.
    """
    # Extracting the file ID from the Google Drive URL
    file_id = url.split("/d/")[1].split("/")[0]
    download_url = f"https://drive.google.com/uc?id={file_id}"
    
    # Load the data into a pandas DataFrame
    return pd.read_csv(download_url)

# Assign URL to a variable
url = 'https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link'

# Loading the data using the function and defining the data as the dataframe
df = load_data_from_google_drive(url)

# Display the first few rows
print(df.head())


   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0


# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [4]:
# Assign URL to a variable
url = 'https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link'

# Loading the data using the function and defining the data as the dataframe
df = load_data_from_google_drive(url)

# Display the first few rows of the dataFrame
print(df.head())

   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0


In [5]:
#Display the dataframe with number of rows and columns
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [6]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [7]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [8]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [9]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [10]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [11]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [12]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [13]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [14]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [15]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [16]:
# Find the number of unique seasons
unique_seasons = df['Season'].nunique()

# Print the results as a sentence
print(f"The number of unique seasons in the dataset is: {unique_seasons}")

The number of unique seasons in the dataset is: 32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [17]:
# Find the team with the most wins
most_wins = df['Wteam'].value_counts().idxmax()  # Team with the most wins
wins_count = df['Wteam'].value_counts().max()    # Number of wins for that team

#Display the resuluts as (most wins, team wins)
most_wins, wins_count


(1181, 819)

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [18]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [19]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [20]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [21]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [22]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [23]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [24]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [25]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [26]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [27]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [28]:
# Sort by Season
df_sorted_by_season = df.sort_values(by='Season')

# Sort by Winning Team (Wteam)
df_sorted_by_wteam = df.sort_values(by='Wteam')

# Sort by Winning Score (Wscore)
df_sorted_by_wscore = df.sort_values(by='Wscore')

#Select rows from index 100-200 for the above dataframes
# Subset from df_sorted_by_season
subset_season = df_sorted_by_season.iloc[100:201, [df.columns.get_loc('Season'), df.columns.get_loc('Wteam'), df.columns.get_loc('Wscore')]]

# Subset from df_sorted_by_wteam
subset_wteam = df_sorted_by_wteam.iloc[100:201, [df.columns.get_loc('Season'), df.columns.get_loc('Wteam'), df.columns.get_loc('Wscore')]]

# Subset from df_sorted_by_wscore
subset_wscore = df_sorted_by_wscore.iloc[100:201, [df.columns.get_loc('Season'), df.columns.get_loc('Wteam'), df.columns.get_loc('Wscore')]]



**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [29]:
# Find the row with the highest winning score in subset_wscore
highest_score_row = subset_wscore.loc[subset_wscore['Wscore'].idxmax()]

# Extract the season and winning team
season_highest_score = highest_score_row['Season']
winning_team_highest_score = highest_score_row['Wteam']

season_highest_score, winning_team_highest_score

#Display the results in a proper context and make it understandable:
print("\nGame with the highest winning score:")
print(f"Season: {season_highest_score}, Winning Team: {winning_team_highest_score}")


Game with the highest winning score:
Season: 2006, Winning Team: 1416


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [30]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [31]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [32]:
# Create a new column 'ScoreDifference' which is the absolute difference between winning score (Wscore) and losing score (Lscore)
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])
print("New column 'ScoreDifference' has been added to the DataFrame.")

#Filter the dataframe in order to calculate the average 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()
print(f"Average ScoreDifference: {average_score_difference:.2f}")

# Filter the DataFrame to only include games with 'ScoreDifference' greater than the average
filtered_df = df[df['ScoreDifference'] > average_score_difference]
print(f"\nFiltered DataFrame contains {len(filtered_df)} games where 'ScoreDifference' > average.")

# Display the table with the new column
print("\nUpdated DataFrame with 'ScoreDifference':")
print(df.head(10).to_string(index=False))  # Display the first 10 rows as a table

New column 'ScoreDifference' has been added to the DataFrame.
Average ScoreDifference: 12.10

Filtered DataFrame contains 57227 games where 'ScoreDifference' > average.

Updated DataFrame with 'ScoreDifference':
 Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference
   1985      20   1228      81   1328      64    N      0               17
   1985      25   1106      77   1354      70    H      0                7
   1985      25   1112      63   1223      56    H      0                7
   1985      25   1165      70   1432      54    H      0               16
   1985      25   1192      86   1447      74    H      0               12
   1985      25   1218      79   1337      78    H      0                1
   1985      25   1228      64   1226      44    N      0               20
   1985      25   1242      58   1268      56    N      0                2
   1985      25   1260      98   1133      80    H      0               18
   1985      25   1305      97   1424 

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [33]:
# Find the row with the highest 'ScoreDifference'
highest_score_diff_row = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

# Extract season and teams involved
season_highest_diff = highest_score_diff_row['Season']
winning_team = highest_score_diff_row['Wteam']
losing_team = highest_score_diff_row['Lteam']
highest_score_diff = highest_score_diff_row['ScoreDifference']

#Display the results in a sentence
print(f"\nGame with the highest ScoreDifference:")
print(f"Season: {season_highest_diff}, Winning Team: {winning_team}, Losing Team: {losing_team}, ScoreDifference: {highest_score_diff}")



Game with the highest ScoreDifference:
Season: 1996, Winning Team: 1409, Losing Team: 1341, ScoreDifference: 91


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [34]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [35]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [36]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [37]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [38]:
# Group by season and calculate the average winning score
average_winning_score = df.groupby('Season')['Wscore'].mean()

#Display the season and the average winning score for that season
print("Average Winning Score per Season:")
print(average_winning_score.to_string())


Average Winning Score per Season:
Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [39]:
# Group by winning team and calculate the maximum winning score
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()

#Display the results of the maximum winning score per team and their winning score
print("\nMaximum Winning Score per Team:")
print(max_winning_score_by_team.to_string())



Maximum Winning Score per Team:
Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
1106    151
1107    108
1108    120
1109    137
1110    128
1111    111
1112    133
1113    123
1114    125
1115    106
1116    166
1117    116
1118     80
1119    104
1120    120
1121    103
1122    115
1123    115
1124    116
1125    107
1126    112
1127     94
1128     92
1129    116
1130    123
1131    110
1132    136
1133    139
1134    101
1135    115
1136    103
1137    116
1138    110
1139    144
1140    128
1141    115
1142    122
1143    121
1144    104
1145    108
1146    112
1147    124
1148    129
1149    114
1150    117
1151    125
1152    128
1153    116
1154    112
1155    120
1156    121
1157    115
1158    116
1159    111
1160    132
1161    108
1162    108
1163    130
1164    110
1165    108
1166    111
1167     98
1168    117
1169    123
1170    110
1171    105
1172    110
1173    118
1174    113
1175    124
1176    103
1177    125
1178    115
1179    109
1180    110
1

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [40]:
# Group by both season and winning team, calculate average winning score
average_score_by_season_team = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Reset index to make it easier to find the team with the highest average score
average_score_by_season_team = average_score_by_season_team.reset_index()

# Find the team with the highest average score for each season
highest_avg_score_team = average_score_by_season_team.loc[average_score_by_season_team.groupby('Season')['Wscore'].idxmax()]

#Display the results as an understandable table for every season
print("\nTeam with Highest Average Winning Score per Season:")
print(highest_avg_score_team.to_string(index=False))



Team with Highest Average Winning Score per Season:
 Season  Wteam     Wscore
   1985   1328  92.800000
   1986   1109  91.200000
   1987   1380  95.875000
   1988   1258 111.750000
   1989   1258 117.315789
   1990   1258 126.347826
   1991   1380 112.312500
   1992   1380  99.642857
   1993   1380 101.875000
   1994   1380 106.583333
   1995   1206 102.833333
   1996   1206 103.600000
   1997   1254  94.142857
   1998   1395 101.153846
   1999   1317  95.750000
   2000   1395  95.647059
   2001   1395  99.315789
   2002   1242  92.888889
   2003   1395  88.222222
   2004   1194  92.142857
   2005   1311  91.333333
   2006   1190  89.533333
   2007   1440  99.400000
   2008   1440  94.500000
   2009   1377  96.200000
   2010   1440  95.000000
   2011   1449  90.130435
   2012   1370  87.400000
   2013   1377  86.500000
   2014   1322  92.600000
   2015   1322  90.733333
   2016   1146  97.142857


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [41]:
# Define win counts by grouping by 'Season' and 'Wteam', and counting the number of games
win_counts = df.groupby(['Season', 'Wteam']).size().reset_index(name='WinCount')

# Display the first 5 rows from the top
print("Top 5 Rows:")
print(win_counts.head().to_string(index=False))

# Display the last 5 rows from the bottom
print("\nBottom 5 Rows:")
print(win_counts.tail().to_string(index=False))



Top 5 Rows:
 Season  Wteam  WinCount
   1985   1102         5
   1985   1103         9
   1985   1104        21
   1985   1106        10
   1985   1108        19

Bottom 5 Rows:
 Season  Wteam  WinCount
   2016   1460        20
   2016   1461        12
   2016   1462        27
   2016   1463        21
   2016   1464         9


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [42]:
# Find the team with the most wins in each season
most_wins_team_per_season = win_counts.loc[win_counts.groupby('Season')['WinCount'].idxmax()]

#Display every season with its associated winning team and the number of wins
print("\nTeam with Most Wins per Season:")
print(most_wins_team_per_season.to_string(index=False))



Team with Most Wins per Season:
 Season  Wteam  WinCount
   1985   1385        27
   1986   1181        32
   1987   1424        33
   1988   1112        31
   1989   1328        28
   1990   1247        29
   1991   1116        30
   1992   1181        28
   1993   1231        28
   1994   1163        27
   1995   1116        26
   1996   1269        31
   1997   1242        31
   1998   1242        33
   1999   1181        32
   2000   1409        29
   2001   1181        29
   2002   1153        30
   2003   1166        29
   2004   1390        29
   2005   1228        32
   2006   1181        30
   2007   1242        30
   2008   1272        33
   2009   1272        31
   2010   1242        32
   2011   1242        32
   2012   1246        32
   2013   1211        30
   2014   1455        33
   2015   1246        34
   2016   1242        29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [43]:
# Group the DataFrame by losing team and calculate the average losing score
average_losing_score = df.groupby('Lteam')['Lscore'].mean()

print("Average Losing Score per Team:")
print(average_losing_score.to_string())

# Group the DataFrame by winning team and calculate the average winning score
average_winning_score = df.groupby('Wteam')['Wscore'].mean()

print("\nAverage Winning Score per Team:")
print(average_winning_score.to_string())

# Compare the average losing and winning scores by merging the two results on team 
comparison_df = pd.DataFrame({
    'AverageLosingScore': average_losing_score,
    'AverageWinningScore': average_winning_score
}).dropna()  # Drop teams that appear only in one group

# Find teams where the average losing score is greater than the average winning score
higher_losing_score_teams = comparison_df[comparison_df['AverageLosingScore'] > comparison_df['AverageWinningScore']]

print("\nTeams with Higher Average Losing Score than Winning Score:")
print(higher_losing_score_teams)


# Find teams where losing score is close to winning score (within 5 points)
close_losing_score_teams = comparison_df[
    abs(comparison_df['AverageLosingScore'] - comparison_df['AverageWinningScore']) <= 5
]

# Print the results of teams with losing scores close to their winning scores
print("\nTeams with Losing Scores Close to Winning Scores:")
print(close_losing_score_teams)

Average Losing Score per Team:
Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
1106    66.284165
1107    60.831418
1108    65.429825
1109    79.535433
1110    62.217659
1111    66.707527
1112    70.055319
1113    66.475983
1114    64.728606
1115    57.530831
1116    67.219718
1117    64.657080
1118    64.190476
1119    60.003419
1120    66.269978
1121    65.078740
1122    67.954936
1123    62.585082
1124    65.868304
1125    68.454054
1126    61.830846
1127    58.003509
1128    60.153846
1129    63.602597
1130    64.616558
1131    62.898104
1132    62.498911
1133    62.018908
1134    64.420000
1135    61.935238
1136    59.802469
1137    63.961340
1138    65.140704
1139    63.865103
1140    68.307692
1141    63.852995
1142    64.953804
1143    65.958869
1144    63.331492
1145    64.651663
1146    64.048780
1147    66.796976
1148    65.123752
1149    64.607143
1150    66.436321
1151    67.072539
1152    64.968182
1153    64.132075
1154    6

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [44]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [45]:
# Create a new column 'HighScoringGame'
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

# Display the first 10 rows of the DataFrame
print("DataFrame with 'HighScoringGame' column:")
print(df.head(10).to_string(index=False))


DataFrame with 'HighScoringGame' column:
 Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference HighScoringGame
   1985      20   1228      81   1328      64    N      0               17              No
   1985      25   1106      77   1354      70    H      0                7              No
   1985      25   1112      63   1223      56    H      0                7              No
   1985      25   1165      70   1432      54    H      0               16              No
   1985      25   1192      86   1447      74    H      0               12              No
   1985      25   1218      79   1337      78    H      0                1              No
   1985      25   1228      64   1226      44    N      0               20              No
   1985      25   1242      58   1268      56    N      0                2              No
   1985      25   1260      98   1133      80    H      0               18              No
   1985      25   1305      97   1424      89    

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [46]:
# Dictionary to keep track of the total number of games for each team
team_games = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Increment games for the winning team
    if row['Wteam'] in team_games:
        team_games[row['Wteam']] += 1
    else:
        team_games[row['Wteam']] = 1
    
    # Increment games for the losing team
    if row['Lteam'] in team_games:
        team_games[row['Lteam']] += 1
    else:
        team_games[row['Lteam']] = 1

# Display the total number of games played by each team
print("\nTotal games played by each team:")
for team, games in list(team_games.items())[:10]:  # Displaying a sample
    print(f"Team {team}: {games} games")



Total games played by each team:
Team 1228: 992 games
Team 1328: 968 games
Team 1106: 855 games
Team 1354: 906 games
Team 1112: 981 games
Team 1223: 363 games
Team 1165: 833 games
Team 1432: 69 games
Team 1192: 908 games
Team 1447: 903 games


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [47]:
# Dictionary to track the highest score difference for each season
highest_diff_games = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']
    
    # If the season is not in the dictionary or if this game has a higher score difference
    if season not in highest_diff_games or score_diff > highest_diff_games[season]['ScoreDifference']:
        highest_diff_games[season] = {
            'Season': season,
            'WinningTeam': row['Wteam'],
            'LosingTeam': row['Lteam'],
            'ScoreDifference': score_diff
        }

# Display the game with the highest score difference for each season
print("\nGame with the highest score difference per season:")
for season, game_info in highest_diff_games.items():
    print(f"Season {season}: Winning Team {game_info['WinningTeam']}, Losing Team {game_info['LosingTeam']}, "
          f"Score Difference {game_info['ScoreDifference']}")



Game with the highest score difference per season:
Season 1985: Winning Team 1361, Losing Team 1288, Score Difference 60
Season 1986: Winning Team 1314, Losing Team 1264, Score Difference 84
Season 1987: Winning Team 1155, Losing Team 1118, Score Difference 73
Season 1988: Winning Team 1328, Losing Team 1147, Score Difference 68
Season 1989: Winning Team 1242, Losing Team 1135, Score Difference 70
Season 1990: Winning Team 1181, Losing Team 1217, Score Difference 76
Season 1991: Winning Team 1163, Losing Team 1148, Score Difference 68
Season 1992: Winning Team 1116, Losing Team 1126, Score Difference 82
Season 1993: Winning Team 1328, Losing Team 1197, Score Difference 81
Season 1994: Winning Team 1228, Losing Team 1152, Score Difference 69
Season 1995: Winning Team 1246, Losing Team 1404, Score Difference 74
Season 1996: Winning Team 1409, Losing Team 1341, Score Difference 91
Season 1997: Winning Team 1278, Losing Team 1106, Score Difference 80
Season 1998: Winning Team 1395, Losing

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [48]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [49]:
# Count the occurrences of each team in Wteam and Lteam
winning_team_counts = df['Wteam'].value_counts()
losing_team_counts = df['Lteam'].value_counts()

# Combine the counts (fill missing teams with 0)
total_games_played = winning_team_counts.add(losing_team_counts, fill_value=0).astype(int)

# Display the total number of games played by each team
print("Total games played by each team (Vectorized):")
print(total_games_played.sort_values(ascending=False).head(10))  # Display the top 10 teams


Total games played by each team (Vectorized):
1181    1013
1314    1010
1246     999
1257     994
1393     993
1242     993
1228     992
1272     992
1437     983
1400     982
Name: count, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [50]:
# Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Group by season and find the row with the maximum score difference
highest_diff_games_per_season = df.loc[df.groupby('Season')['ScoreDifference'].idxmax()]

# Display the results
print("\nGame with the highest score difference for each season:")
print(highest_diff_games_per_season[['Season', 'Wteam', 'Lteam', 'ScoreDifference']])



Game with the highest score difference for each season:
        Season  Wteam  Lteam  ScoreDifference
236       1985   1361   1288               60
4731      1986   1314   1264               84
8240      1987   1155   1118               73
12046     1988   1328   1147               68
16677     1989   1242   1135               70
19502     1990   1181   1217               76
25161     1991   1163   1148               68
27997     1992   1116   1126               82
33858     1993   1328   1197               81
36404     1994   1228   1152               69
39858     1995   1246   1404               74
44653     1996   1409   1341               91
49033     1997   1278   1106               80
52600     1998   1395   1410               66
57618     1999   1268   1317               75
60616     2000   1261   1212               75
65999     2001   1196   1197               75
70737     2002   1328   1183               59
74791     2003   1403   1311               72
79181     2004   1328  

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [51]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [52]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [53]:
type(df['Wscore'])

pandas.core.series.Series

In [54]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [55]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [56]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [57]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y